flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArrayPositionFunction.java

```diff
@@ -0,0 +1,91 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.runtime.functions.scalar;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.table.api.DataTypes;
+import org.apache.flink.table.data.ArrayData;
+import org.apache.flink.table.functions.BuiltInFunctionDefinitions;
+import org.apache.flink.table.functions.FunctionContext;
+import org.apache.flink.table.functions.SpecializedFunction.ExpressionEvaluator;
+import org.apache.flink.table.functions.SpecializedFunction.SpecializedContext;
+import org.apache.flink.table.types.CollectionDataType;
+import org.apache.flink.table.types.DataType;
+import org.apache.flink.util.FlinkRuntimeException;
+
+import javax.annotation.Nullable;
+
+import java.lang.invoke.MethodHandle;
+
+import static org.apache.flink.table.api.Expressions.$;
+
+/** Implementation of {@link BuiltInFunctionDefinitions#ARRAY_POSITION}. */
+@Internal
+public class ArrayPositionFunction extends BuiltInScalarFunction {
+
+    private final ArrayData.ElementGetter elementGetter;
+    private final ExpressionEvaluator equalityEvaluator;
+    private transient MethodHandle equalityHandle;
+
+    public ArrayPositionFunction(SpecializedContext context) {
+        super(BuiltInFunctionDefinitions.ARRAY_POSITION, context);
+        final DataType elementDataType =
```

Comment:
Can we only get the `needleDataType`, and use `needleDataType` as the `elementDataType` just like what [`ArrayContainsFunction`](https://github.com/apache/flink/blob/master/flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArrayContainsFunction.java#L47) does?

Reply:
because there is a bug in array_contains https://issues.apache.org/jira/browse/FLINK-31377 
so array_remove i supported is also using it  https://github.com/apache/flink/blob/master/flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArrayRemoveFunction.java

Reply:
Thanks for clarify.  Another question, IIUC, the `needleDataType` will be casted to `elementDataType`, so can we get `elementDataType` and use `elementDataType` as the `needleDataType `?

Reply:
the cast will insert, and the elementDataType and needleDataType is not identical when handling the nullable
because " a NOT NULL type can be stored in NULL type but not vice versa."

Reply:
so can not, you can look this pr https://github.com/apache/flink/pull/22143

Reply:
I see. But from the pr #22143 
```
final DataType elementDataType =
                ((CollectionDataType) context.getCallContext().getArgumentDataTypes().get(0))
                        .getElementDataType();
        elementGetter = ArrayData.createElementGetter(elementDataType.getLogicalType());
        equalityEvaluator =
                context.createEvaluator(
                        $("element").isEqual($("needle")),
                        DataTypes.BOOLEAN(),
                        DataTypes.FIELD("element", elementDataType.notNull().toInternal()),
                        DataTypes.FIELD("needle", elementDataType.notNull().toInternal()));
```
Isn't it also get `elementDataType` and use `elementDataType` as `needleDataType`?